sycl/doc/SYCL_compiler_and_runtime_design.md

```diff
@@ -0,0 +1,262 @@
+SYCL source. The result of compilation is a set of so-called "fat objects" -
+one fat object per SYCL source file. A fat object contains compiled host code
+and a number of compiled device code instances - one per each target. Fat
+objects can be linked into "fat binary".
+
+SYCL sources can be also compiled as a regular C++ code, in this mode there is
+no "device part" of the code - everything is executed on the host.
+
+Device compiler is further split into the following major components:
+- **Front-end** - parses input source, outlines "device part" of the code,
+applies additional restrictions on the device code (e.g. no exceptions or
+virtual calls), generated LLVM IR for the device code only and "integration
+header" which provides information like kernel name, parameters order and data
+type for the runtime library.
+- **Middle-end** - transforms the initial LLVM IR* to get consumed by the
+back-end. Today middle-end transformations include just a couple of passes:
+  - OpenCL C++* to SPIR-V* built-in function names mapper
+  - Address space handling pass
+  - TBD: potentially the middle-end optimizer can run any LLVM IR
+  transformation with only one limitation: back-end compiler should be able to
+  handle transformed LLVM IR.
+  - Optionally: LLVM IR -> SPIR-V translator.
+- **Back-end** - produces native "device" code in ahead-of-time compilation
+mode.
+
+### SYCL support in the driver
+SYCL offload support in the driver is based on the clang driver concepts and
+defines:
+- target triple and a native tool chain for each target (including "virtual"
+targets like SPIR-V).
+- SYCL offload action based on generic offload action
+
+#### Enable SYCL offload
+To enable compilation with SYCL specification conformance, a special option
+must be passed to the clang driver:
+
+`-fsycl`
+
+With this option specified, the driver will invoke the host SYCL compiler and a
+number of device compilers for targets specified in the `-fsycl-targets`
+option.  If this option is not specified, then single SPIR-V target is assumed,
+and single device compiler for this target is invoked.
+
+#### Ahead of time (AOT) compilation
+
+Ahead-of-time compilation is the process of invoking the back-end at compile
+time to produce the final binary, as opposed to just-in-time (JIT) compilation
+when final code generation is deferred until application runtime time.
+
+AOT compilation reduces application execution time by skipping JIT compilation
+and final device code can be tested before deploy.
+
+JIT compilation provides portability of device code and target specific
+optimizations.
+
+#### List of native targets
+The ahead-of-time compilation mode implies that there must be a way to specify
+a set of target architectures for which to compile device code. By default the
+compiler generates SPIR-V and OpenCL device JIT compiler produces native target
+binary.
+
+There are existing options for OpenMP* offload:
```

Comment:
I have the feeling you mean that you get the inspiration from OpenMP as an existing example, but it is not clear enough.
This suggests more that we can also have OpenMP as the same time as SYCL. But perhaps it just happens to work like that too? It would be great to be able to use OpenMP and SYCL in the same program, for example for HPC applications. Of course not OpenMP inside a SYCL kernel, except on a host queue.